ui_framework/components/button/button.js

```diff
@@ -0,0 +1,135 @@
+import React, {
+  PropTypes,
+} from 'react';
+
+import classNames from 'classnames';
+import keyMirror from 'keymirror';
+
+import { KuiLoadingButtonIcon } from './button_icon';
+
+const KuiButton = props => {
+  if (props.isSubmit) {
+    // input is a void element tag and can't have children.
+    if ((props.children && typeof props.children !== 'string') || props.icon) {
+      throw new Error(
+        `KuiButton with isSubmit prop set to true can only accept string children, and can't
+        display icons. This is because input is a void element and can't contain children.`
+      );
+    }
+  }
+
+  function onClick() {
+    // Don't even trigger the onClick handler if we're disabled.
+    if (props.isDisabled) {
+      return;
+    }
+    props.onClick(props.data);
+  }
+
+  const icon =
+    props.isLoading
+    ? <KuiLoadingButtonIcon />
+    : props.icon;
+
+  const className = classNames('kuiButton', props.className, {
+    'kuiButton--iconText': icon,
+  });
+
+  let wrappedChildren;
+
+  if (props.children) {
+    // We need to wrap the text so that the icon's :first-child etc. seudo-selectors get applied
+    // correctly.
+    wrappedChildren = (
+      <span>{props.children}</span>
+    );
+  }
+
+  const elementType =
+    props.isSubmit
+    ? 'input'
+    : props.href
+    ? 'a'
+    : 'button';
+
+  const children =
+    props.isSubmit
+    ? null
+    : props.isIconOnRight
+    ? (
+      <span>
+        {wrappedChildren}
+        {icon}
+      </span>
+    ) : (
+      <span>
+        {icon}
+        {wrappedChildren}
+      </span>
+    );
+
+  return React.createElement(elementType, {
+    'data-test-subj': props.testSubject,
+    className,
+    href: props.href,
+    target: props.target,
+    type: props.isSubmit ? 'submit' : null,
+    disabled: props.isDisabled,
+    // onClick is optional, so don't even call it if it's not passed in.
+    onClick: props.onClick ? onClick : null,
```

Comment:
This is a bit weird to me. It's checking `props.onClick` but passing in `onClick`. I see that thing is using `props.onClick` internally, but it makes this code less understandable. Maybe do the `isDisabled` check here, then if not disabled pass in `onClick`?

Also, as an api having `onClick` being potentially `null` is a bad api for underlying components. Better to give it a noop function? (e.g. `() => {}`, basically just an empty function that doens't do anything when called)